SwrveSDK/src/main/java/com/swrve/sdk/SwrveBase.java

```diff
@@ -179,15 +184,35 @@ protected T init(final Activity activity) throws IllegalArgumentException {
                 // Add custom message listener
                 if (messageListener == null) {
                     setMessageListener(new ISwrveMessageListener() {
-                        public void onMessage(final SwrveMessage message, boolean firstTime) {
-                            if (SwrveBase.this.activityContext != null) {
-                                final Activity activity = SwrveBase.this.activityContext.get();
-                                if (activity == null) {
-                                    SwrveLogger.e(LOG_TAG, "Can't display a message with a non-Activity context");
+                        public void onMessage(final SwrveMessage message) {
+                            // Start a Conversation activity to display the campaign
+                            if (SwrveBase.this.context != null) {
+                                final Context ctx = SwrveBase.this.context.get();
+                                if (ctx == null) {
+                                    SwrveLogger.e(LOG_TAG, "Can't display a conversation without a context");
                                     return;
                                 }
-                                // Run code on the UI thread
-                                activity.runOnUiThread(new DisplayMessageRunnable(SwrveBase.this, activity, message, firstTime));
+
+                                Intent intent = new Intent(ctx, SwrveInAppMessageActivity.class);
+                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                                intent.putExtra("inAppMessage", message);
+                                intent.putExtra("hideToolbar", config.isHideToolbar());
+                                intent.putExtra("minSampleSize", config.getMinSampleSize());
```

Comment:
Ok, am going back on this. Making everything serializable is not good. (Should be parcelable anyway, but we shouldn't do that either). IAM should be dependent on the singleton. The whole sdk, to my knowledge, is anyways.